src/event.rs

```diff
@@ -1,88 +1,99 @@
+// Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
+//
+// This program is licensed to you under the Apache License Version 2.0,
+// and you may not use this file except in compliance with the Apache License Version 2.0.
+// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the Apache License Version 2.0 is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+
 use derive_builder::Builder;
 use serde::{Deserialize, Serialize};
-use serde_json::json;
-use serde_json::Value;
-
-#[derive(Serialize, Deserialize)]
-pub struct Event {
-    pub true_timestamp: i64,
-    pub context: Vec<SelfDescribingJson>,
-}
+use serde_json::{Value, json};
+use uuid::Uuid;
 
-impl Event {
-    pub fn set_context(&mut self, contexts: Vec<SelfDescribingJson>) -> &mut Event {
-        self.context = contexts;
-        self
-    }
+use crate::payload::{EventType, SelfDescribingEventData, PayloadBuilder, Payload, SelfDescribingJson};
 
-    pub fn set_true_timestamp(&mut self, timestamps: i64) -> &mut Event {
-        self.true_timestamp = timestamps;
-        self
-    }
+/// Trait implemented by event types that enables the tracker to build their payload to be sent to the Collector.
+pub trait EventBuildable {
+    fn build_payload(self, payload_builder: PayloadBuilder) -> Payload;
 }
 
-#[derive(Serialize, Deserialize)]
+/// Event to track custom information that does not fit into the out-of-the box events.
+///
+/// Self-describing events are a [data structure based on JSON Schemas](https://docs.snowplowanalytics.com/docs/understanding-tracking-design/understanding-schemas-and-validation/) and can have arbitrarily many fields.
```

Comment:
I wrote some rambling about this docstring, but the main point I want to make is that I think the explanation of creating a custom schema should exist on the Snowplow docs rather than a docstring of a data structure, but I can't really find anything suitable to link out to (it seems like a page that really should exist).

If at some point in the future (or if you know of one that currently exists) there's a decent page to link out to, I would say this is worth cutting down the how-to:

" ... To define your own custom self-describing event follow [some wonderful guide that explains it all]() ..."

But for now, I think this is fine.

Reply:
I agree with this, the docstring is not a good place to explain how to define self-describing event schemas. I think I'll just remove that sentence, it's quite likely to get outdated relatively fast and I don't think we'll be able to keep the info up to date. Once we have the wonderful guide, we can add a link back.